### PR TITLE
PR To support DbDataSource for Postgresql

### LINKED ISSUE #1813: AspNetCore.HealthChecks.NpgSql (Postgres) health check and Managed Identity on Azure
The current health check for Postgres (AspNetCore.HealthChecks.NpgSql, version 6.0.2) does not support Azure Managed Identity tokens as far as I can see.

Are there any workarounds for this, or any plans on adding support for Managed Identity tokens?

Using Managed Identity tokens is increasingly normal to use, so supporting this scenario would be great.

### PR DESCRIPTION
**Pull request by michaelmairegger**
This PR adds DbDataSource support for postgresql to allow password rotation using UsePeriodicPasswordProvider

This PR allows the following:

```
var dataSourceBuilder = new NpgsqlDataSourceBuilder(connectionString);

dataSourceBuilder.UsePeriodicPasswordProvider(async (settings, cancellationToken) =>
{
    var sqlServerTokenProvider = new DefaultAzureCredential();
    var tokenAsync = await sqlServerTokenProvider.GetTokenAsync(new TokenRequestContext(scopes: new[] { "https://ossrdbms-aad.database.windows.net/.default" }), cancellationToken);
    return tokenAsync.Token;
},
TimeSpan.FromMinutes(55), // Interval for refreshing the token
TimeSpan.FromSeconds(5));
        
var dbDataSource = dataSourceBuilder.Build();


builder.Services.AddHealthChecks()
    .AddNpgSql(e => dbDataSource);
```

closes #1813

### CODE REVIEW

File: src/HealthChecks.NpgSql/NpgSqlHealthCheckOptions.cs [hunk 1]
```diff
@@ -10,9 +10,9 @@ namespace HealthChecks.NpgSql;
 public class NpgSqlHealthCheckOptions
 {
     /// <summary>
-    /// The Postgres connection string to be used.
+    /// The Postgres data source to be used.
     /// </summary>
-    public string ConnectionString { get; set; } = null!;
```
Comment (rob-baldwin) [inline src/HealthChecks.NpgSql/NpgSqlHealthCheckOptions.cs:15]: Removing this property will make this a breaking change.  Could you alter this PR to keep the `ConnectionString` property please?

Comment (michaelmairegger) [reply src/HealthChecks.NpgSql/NpgSqlHealthCheckOptions.cs:15]: Ok, well thanks. Settings ConnectionString should set the DbSource

File: src/HealthChecks.NpgSql/NpgSqlHealthCheckOptions.cs [hunk 2]
```diff
@@ -11,8 +11,18 @@ public class NpgSqlHealthCheckOptions
 {
     /// <summary>
     /// The Postgres connection string to be used.
+    /// The Postgres data source to be used.
```
Comment (sungam3r) [inline src/HealthChecks.NpgSql/NpgSqlHealthCheckOptions.cs:14]:
```suggestion
```

File: src/HealthChecks.NpgSql/DependencyInjection/NpgSqlHealthCheckBuilderExtensions.cs [hunk 3]
```diff
@@ -65,15 +65,28 @@ public static class NpgSqlHealthCheckBuilderExtensions
         IEnumerable<string>? tags = default,
         TimeSpan? timeout = default)
     {
-        Guard.ThrowIfNull(connectionStringFactory);
+        return builder.AddNpgSql(e => new NpgsqlDataSourceBuilder(connectionStringFactory(e)).Build(), healthQuery, configure, name, failureStatus, tags, timeout);
```
Comment (sungam3r) [inline src/HealthChecks.NpgSql/DependencyInjection/NpgSqlHealthCheckBuilderExtensions.cs:68]:
```suggestion
        return builder.AddNpgSql(sp => new NpgsqlDataSourceBuilder(connectionStringFactory(sp)).Build(), healthQuery, configure, name, failureStatus, tags, timeout);
```

File: src/HealthChecks.NpgSql/DependencyInjection/NpgSqlHealthCheckBuilderExtensions.cs [hunk 4]
```diff
@@ -65,15 +65,28 @@ public static class NpgSqlHealthCheckBuilderExtensions
         IEnumerable<string>? tags = default,
         TimeSpan? timeout = default)
     {
-        Guard.ThrowIfNull(connectionStringFactory);
+        return builder.AddNpgSql(sp => new NpgsqlDataSourceBuilder(connectionStringFactory(sp)).Build(), healthQuery, configure, name, failureStatus, tags, timeout);
+    }
+
+    public static IHealthChecksBuilder AddNpgSql(
```
Comment (sungam3r) [inline src/HealthChecks.NpgSql/DependencyInjection/NpgSqlHealthCheckBuilderExtensions.cs:71]:
```suggestion
    /// <summary>
    /// Add a health check for Postgres databases.
    /// </summary>
    /// <param name="builder">The <see cref="IHealthChecksBuilder"/>.</param>
    /// <param name="dbDataSourceFactory">A factory to build the NpgsqlDataSource to use.</param>
    /// <param name="healthQuery">The query to be used in check.</param>
    /// <param name="configure">An optional action to allow additional Npgsql specific configuration.</param>
    /// <param name="name">The health check name. Optional. If <c>null</c> the type name 'npgsql' will be used for the name.</param>
    /// <param name="failureStatus">
    /// The <see cref="HealthStatus"/> that should be reported when the health check fails. Optional. If <c>null</c> then
    /// the default status of <see cref="HealthStatus.Unhealthy"/> will be reported.
    /// </param>
    /// <param name="tags">A list of tags that can be used to filter sets of health checks. Optional.</param>
    /// <param name="timeout">An optional <see cref="TimeSpan"/> representing the timeout of the check.</param>
    /// <returns>The specified <paramref name="builder"/>.</returns>
    public static IHealthChecksBuilder AddNpgSql(
```

File: src/HealthChecks.NpgSql/NpgSqlHealthCheckOptions.cs [hunk 5]
```diff
@@ -12,7 +12,16 @@ public class NpgSqlHealthCheckOptions
     /// <summary>
     /// The Postgres connection string to be used.
```
Comment (sungam3r) [inline src/HealthChecks.NpgSql/NpgSqlHealthCheckOptions.cs:13]:
```suggestion
    /// The Postgres connection string to be used.
    /// Use <see cref="DataSource"/> property for advanced configuration.
```

**Review by sungam3r — APPROVED** [review]

**Comment by sungam3r** [comment]
@michaelmairegger Thanks.

**Comment by michaelmairegger** [comment]
@sungam3r Oops, my bad. Thanks

**Pull request merged**